spec/create_new_milestone_action_spec.rb

```diff
@@ -0,0 +1,156 @@
+        expect(glithub_helper).to receive(:create_milestone).with(
+          anything,
+          anything,
+          anything,
+          anything,
+          default_code_freeze_days,
+          anything
+        )
+        run_described_fastlane_action(mock_params)
+      end
+
+      it 'uses default value when neither `GHHELPER_MILESTONE_DURATION` environment variable nor parameter `:milestone_duration` is present' do
+        default_milestone_duration = 14
+        mock_params[:milestone_duration] = nil
+        ENV['GHHELPER_MILESTONE_DURATION'] = nil
+        expect(glithub_helper).to receive(:create_milestone).with(
+          anything,
+          anything,
+          anything,
+          default_milestone_duration,
+          anything,
+          anything
+        )
+        run_described_fastlane_action(mock_params)
+      end
+    end
+  end
+
+  describe 'get_last_milestone' do
+    let(:test_repository) { 'test-repository' }
+    let(:test_milestone) { { title: '10.1', due_on: '2022-10-31T07:00:00Z' } }
+    let(:mock_params) do
+      {
+        repository: test_repository,
+        need_appstore_submission: false,
+        github_token: 'Test-GithubToken-1234'
+      }
+    end
+    let(:client) do
+      instance_double(
+        Octokit::Client,
+        list_milestones: [test_milestone],
+        create_milestone: nil,
+        user: instance_double('User', name: 'test'),
+        'auto_paginate=': nil
+      )
+    end
+
+    it 'properly passes the repository all the way down to the Octokit::Client to list the existing milestones' do
+      allow(Octokit::Client).to receive(:new).and_return(client)
+      expect(client).to receive(:list_milestones).with(test_repository, { state: 'open' })
+      run_described_fastlane_action(mock_params)
+    end
+  end
+
+  describe 'create_milestone' do
+    let(:test_repository) { 'test-repository' }
+    let(:test_milestone_number) { '10.2' }
+    let(:test_milestone) { { title: '10.1', due_on: '2022-10-31T07:00:00Z' } }
+    let(:mock_params) do
+      {
+        repository: test_repository,
+        need_appstore_submission: false,
+        github_token: 'Test-GithubToken-1234'
+      }
+    end
+    let(:client) do
+      instance_double(
+        Octokit::Client,
+        list_milestones: [test_milestone],
+        create_milestone: nil,
+        user: instance_double('User', name: 'test'),
+        'auto_paginate=': nil
+      )
+    end
+
+    it 'properly passes the parameters all the way down to the Octokit::Client' do
+      comment = "Code freeze: November 14, 2022\nApp Store submission: November 28, 2022\nRelease: November 28, 2022\n"
+      options = { due_on: '2022-11-14T12:00:00Z', description: comment }
+      allow(Octokit::Client).to receive(:new).and_return(client)
+      expect(client).to receive(:create_milestone).with(test_repository, test_milestone_number, options)
+      run_described_fastlane_action(mock_params)
+    end
```

Comment:
This is the really useful unit test we need when testing the `CreateNewMilestoneAction` as a black box: pass some input parameters, and test that the `create_milestone` method of the `Octokit::Client` is called with the correct and expected parameters and final dates.

Except that there are so many more possible inputs to cover using that test case, especially:
 - Test that the computed dates (in `due_on` and `description`) are correct and the computation of the new dates were based on the right milestone (i.e. when there are multiple milestones returned by `list_milestones`, with different `due_on` dates, ensure the final dates for the new milestone is still correct)
 - Test the case where no milestones were returned
 - Test the case when we pass different offsets
 - Test error cases when we pass inconsistent or invalid offsets, if relevant
 - Test the case when the last milestone is just before a DST change, to ensure our date math takes that into account
 - …

Those are the use cases we need to cover, and they should basically all have the same structure as this `it` unit test above (i.e. their goal is always to `expect(client).to receive(:create_milestone).with(the-correct-params)` given the params passed to `run_described_fastlane_action`), but just with different inputs (different list of returned milestones, existing milestines with different due dates or with missing due dates or whatnot, case of need_appstore_submission true vs false, case of different day offsets, …) and thus different corresponding expected parameters passed to the client.

Reply:
> Shouldn't all this validation about the date offsets, DST, and general due date/description cases that you mention be handled by the Unit tests from GithubHelper instead of the ones from the Action itself? 🤔

That's a good point, indeed.

Thought the implementation of this action (and its split of the code between the action itself and the helper) is apparently currently at an odd place, because some of the date math is done in the action (namely `newmilestone_duedate = (milestone_duedate.to_datetime.next_day(milestone_duration).to_time).utc`), while other parts of the date math is done in `GithubHelper#create_milestone` itself (namely the computation of `submission_date` and `release_date` based on `days_until_submission` and `number_of_days_from_code_freeze_to_release`)…

We should clean that up at some point (the API of that `CreateNewMilestoneAction` is a bit odd anyway and we hope to modernize it at one point to make the `ConfigItem` it declares have better names and API… but that's another story for another time) and once we do, we'd probably have all the date math be done in the action and not in the helper (which would just receive the already-computed dates as input then), which would then make testing the date math in the `create_new_milestone_action_spec` the next sensible thing.

---

That being said, I see you added some tests about this in lines 30–57 since 👍